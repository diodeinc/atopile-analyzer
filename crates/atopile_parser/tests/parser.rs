use atopile_parser::lexer::lex;
use atopile_parser::parser::parse;
use std::fs;

macro_rules! create_parser_test {
    ($name:ident) => {
        #[test]
        fn $name() {
            let input = fs::read_to_string(concat!(
                "tests/resources/corpus/",
                stringify!($name),
                ".ato"
            ))
            .unwrap();

            let normalized_input = input.replace("\r\n", "\n");

            let tokens = lex(&normalized_input).0;
            let output = parse(&tokens);

            for err in output.1.iter() {
                println!("Error: {:?}", err);
                println!("Context tokens:");
                let span = err.span();
                tokens
                    .iter()
                    .enumerate()
                    .filter(|(i, _)| {
                        *i >= span.start.saturating_sub(5) && *i <= span.end.saturating_add(5)
                    })
                    .for_each(|(i, t)| {
                        println!(
                            "{}{:2}: {:?}",
                            if span.into_range().contains(&i) {
                                "-> "
                            } else {
                                "   "
                            },
                            i,
                            t
                        )
                    });
                println!("---");
            }

            insta::assert_debug_snapshot!(output);
        }
    };
}

create_parser_test!(vdivs);
create_parser_test!(resistors);
create_parser_test!(transistors);
create_parser_test!(bma400);
