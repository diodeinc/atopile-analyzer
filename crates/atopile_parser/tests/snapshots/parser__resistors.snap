---
source: crates/atopile_parser/tests/parser.rs
expression: output
---
(
    [
        Spanned(
            DepImport(
                DepImportStmt {
                    name: Spanned(
                        Symbol(
                            "Power",
                        ),
                        7..12,
                    ),
                    from_path: Spanned(
                        "interfaces.ato",
                        18..34,
                    ),
                },
            ),
            0..34,
        ),
        Spanned(
            DepImport(
                DepImportStmt {
                    name: Spanned(
                        Symbol(
                            "I2C",
                        ),
                        42..45,
                    ),
                    from_path: Spanned(
                        "interfaces.ato",
                        51..67,
                    ),
                },
            ),
            35..67,
        ),
        Spanned(
            Block(
                BlockStmt {
                    kind: Spanned(
                        Component,
                        69..78,
                    ),
                    name: Spanned(
                        Symbol(
                            "Resistor",
                        ),
                        79..87,
                    ),
                    parent: None,
                    body: [
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Signal(
                                            Spanned(
                                                "p1",
                                                100..102,
                                            ),
                                        ),
                                        93..102,
                                    ),
                                    right: Spanned(
                                        Pin(
                                            Spanned(
                                                "1",
                                                109..110,
                                            ),
                                        ),
                                        105..110,
                                    ),
                                },
                            ),
                            93..110,
                        ),
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Signal(
                                            Spanned(
                                                "p2",
                                                122..124,
                                            ),
                                        ),
                                        115..124,
                                    ),
                                    right: Spanned(
                                        Pin(
                                            Spanned(
                                                "2",
                                                131..132,
                                            ),
                                        ),
                                        127..132,
                                    ),
                                },
                            ),
                            115..132,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "designator_prefix",
                                                    137..154,
                                                ),
                                            ],
                                        },
                                        137..154,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        String(
                                            Spanned(
                                                "R",
                                                157..160,
                                            ),
                                        ),
                                        157..160,
                                    ),
                                },
                            ),
                            137..160,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "mpn",
                                                    165..168,
                                                ),
                                            ],
                                        },
                                        165..168,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        String(
                                            Spanned(
                                                "generic_resistor",
                                                171..189,
                                            ),
                                        ),
                                        171..189,
                                    ),
                                },
                            ),
                            165..189,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "type",
                                                    194..198,
                                                ),
                                            ],
                                        },
                                        194..198,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        String(
                                            Spanned(
                                                "resistor",
                                                201..211,
                                            ),
                                        ),
                                        201..211,
                                    ),
                                },
                            ),
                            194..211,
                        ),
                        Spanned(
                            Attribute(
                                AttributeStmt {
                                    name: Spanned(
                                        Symbol(
                                            "value",
                                        ),
                                        216..221,
                                    ),
                                    type_info: Spanned(
                                        Symbol(
                                            "resistance",
                                        ),
                                        223..233,
                                    ),
                                },
                            ),
                            216..233,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        " Equations",
                                        239..250,
                                    ),
                                },
                            ),
                            239..250,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        " R = V/I",
                                        255..264,
                                    ),
                                },
                            ),
                            255..264,
                        ),
                        Spanned(
                            Comment(
                                CommentStmt {
                                    comment: Spanned(
                                        " R = V^2/P",
                                        269..280,
                                    ),
                                },
                            ),
                            269..280,
                        ),
                    ],
                },
            ),
            69..282,
        ),
        Spanned(
            Block(
                BlockStmt {
                    kind: Spanned(
                        Module,
                        282..288,
                    ),
                    name: Spanned(
                        Symbol(
                            "I2CPullup",
                        ),
                        289..298,
                    ),
                    parent: None,
                    body: [
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_sda",
                                                    304..309,
                                                ),
                                            ],
                                        },
                                        304..309,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        New(
                                            Spanned(
                                                Symbol(
                                                    "Resistor",
                                                ),
                                                316..324,
                                            ),
                                        ),
                                        312..324,
                                    ),
                                },
                            ),
                            304..324,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_scl",
                                                    329..334,
                                                ),
                                            ],
                                        },
                                        329..334,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        New(
                                            Spanned(
                                                Symbol(
                                                    "Resistor",
                                                ),
                                                341..349,
                                            ),
                                        ),
                                        337..349,
                                    ),
                                },
                            ),
                            329..349,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "power",
                                                    354..359,
                                                ),
                                            ],
                                        },
                                        354..359,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        New(
                                            Spanned(
                                                Symbol(
                                                    "Power",
                                                ),
                                                366..371,
                                            ),
                                        ),
                                        362..371,
                                    ),
                                },
                            ),
                            354..371,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "i2c",
                                                    376..379,
                                                ),
                                            ],
                                        },
                                        376..379,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        New(
                                            Spanned(
                                                Symbol(
                                                    "I2C",
                                                ),
                                                386..389,
                                            ),
                                        ),
                                        382..389,
                                    ),
                                },
                            ),
                            376..389,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_sda",
                                                    395..400,
                                                ),
                                                Spanned(
                                                    "value",
                                                    401..406,
                                                ),
                                            ],
                                        },
                                        395..406,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        Physical(
                                            Spanned(
                                                PhysicalValue {
                                                    value: Spanned(
                                                        "10",
                                                        409..411,
                                                    ),
                                                    unit: Some(
                                                        Spanned(
                                                            "kohm",
                                                            411..415,
                                                        ),
                                                    ),
                                                    tolerance: Some(
                                                        Spanned(
                                                            Bilateral {
                                                                value: Spanned(
                                                                    "20",
                                                                    420..422,
                                                                ),
                                                                unit: None,
                                                            },
                                                            416..423,
                                                        ),
                                                    ),
                                                },
                                                409..423,
                                            ),
                                        ),
                                        409..423,
                                    ),
                                },
                            ),
                            395..423,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_scl",
                                                    428..433,
                                                ),
                                                Spanned(
                                                    "value",
                                                    434..439,
                                                ),
                                            ],
                                        },
                                        428..439,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        Physical(
                                            Spanned(
                                                PhysicalValue {
                                                    value: Spanned(
                                                        "10",
                                                        442..444,
                                                    ),
                                                    unit: Some(
                                                        Spanned(
                                                            "kohm",
                                                            444..448,
                                                        ),
                                                    ),
                                                    tolerance: Some(
                                                        Spanned(
                                                            Bilateral {
                                                                value: Spanned(
                                                                    "20",
                                                                    453..455,
                                                                ),
                                                                unit: None,
                                                            },
                                                            449..456,
                                                        ),
                                                    ),
                                                },
                                                442..456,
                                            ),
                                        ),
                                        442..456,
                                    ),
                                },
                            ),
                            428..456,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_sda",
                                                    462..467,
                                                ),
                                                Spanned(
                                                    "package",
                                                    468..475,
                                                ),
                                            ],
                                        },
                                        462..475,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        String(
                                            Spanned(
                                                "0402",
                                                478..484,
                                            ),
                                        ),
                                        478..484,
                                    ),
                                },
                            ),
                            462..484,
                        ),
                        Spanned(
                            Assign(
                                AssignStmt {
                                    target: Spanned(
                                        PortRef {
                                            parts: [
                                                Spanned(
                                                    "r_scl",
                                                    489..494,
                                                ),
                                                Spanned(
                                                    "package",
                                                    495..502,
                                                ),
                                            ],
                                        },
                                        489..502,
                                    ),
                                    type_info: None,
                                    value: Spanned(
                                        String(
                                            Spanned(
                                                "0402",
                                                505..511,
                                            ),
                                        ),
                                        505..511,
                                    ),
                                },
                            ),
                            489..511,
                        ),
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "i2c",
                                                            517..520,
                                                        ),
                                                        Spanned(
                                                            "sda",
                                                            521..524,
                                                        ),
                                                    ],
                                                },
                                                517..524,
                                            ),
                                        ),
                                        517..524,
                                    ),
                                    right: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "r_sda",
                                                            527..532,
                                                        ),
                                                        Spanned(
                                                            "1",
                                                            533..534,
                                                        ),
                                                    ],
                                                },
                                                527..534,
                                            ),
                                        ),
                                        527..534,
                                    ),
                                },
                            ),
                            517..534,
                        ),
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "r_sda",
                                                            536..541,
                                                        ),
                                                        Spanned(
                                                            "2",
                                                            542..543,
                                                        ),
                                                    ],
                                                },
                                                536..543,
                                            ),
                                        ),
                                        536..543,
                                    ),
                                    right: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "power",
                                                            546..551,
                                                        ),
                                                        Spanned(
                                                            "vcc",
                                                            552..555,
                                                        ),
                                                    ],
                                                },
                                                546..555,
                                            ),
                                        ),
                                        546..555,
                                    ),
                                },
                            ),
                            536..555,
                        ),
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "i2c",
                                                            560..563,
                                                        ),
                                                        Spanned(
                                                            "scl",
                                                            564..567,
                                                        ),
                                                    ],
                                                },
                                                560..567,
                                            ),
                                        ),
                                        560..567,
                                    ),
                                    right: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "r_scl",
                                                            570..575,
                                                        ),
                                                        Spanned(
                                                            "1",
                                                            576..577,
                                                        ),
                                                    ],
                                                },
                                                570..577,
                                            ),
                                        ),
                                        570..577,
                                    ),
                                },
                            ),
                            560..577,
                        ),
                        Spanned(
                            Connect(
                                ConnectStmt {
                                    left: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "r_scl",
                                                            579..584,
                                                        ),
                                                        Spanned(
                                                            "2",
                                                            585..586,
                                                        ),
                                                    ],
                                                },
                                                579..586,
                                            ),
                                        ),
                                        579..586,
                                    ),
                                    right: Spanned(
                                        Port(
                                            Spanned(
                                                PortRef {
                                                    parts: [
                                                        Spanned(
                                                            "power",
                                                            589..594,
                                                        ),
                                                        Spanned(
                                                            "vcc",
                                                            595..598,
                                                        ),
                                                    ],
                                                },
                                                589..598,
                                            ),
                                        ),
                                        589..598,
                                    ),
                                },
                            ),
                            579..598,
                        ),
                    ],
                },
            ),
            282..598,
        ),
    ],
    [],
)
